flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/calls/ScalarOperatorGens.scala

```diff
@@ -445,14 +445,24 @@ object ScalarOperatorGens {
     }
   }
 
+  def generateIsDistinctFrom(
+      ctx: CodeGeneratorContext,
+      left: GeneratedExpression,
+      right: GeneratedExpression)
+    : GeneratedExpression = {
+    generateAnd(
+      generateOr(generateIsNotNull(left), generateIsNotNull(right)),
+      generateIsNotTrue(generateEquals(ctx, left, right)))
+  }
+
   def generateIsNotDistinctFrom(
       ctx: CodeGeneratorContext,
       left: GeneratedExpression,
       right: GeneratedExpression)
     : GeneratedExpression = {
     generateOr(
       generateAnd(generateIsNull(left), generateIsNull(right)),
-      generateEquals(ctx, left, right))
+      generateIsTrue(generateEquals(ctx, left, right)))
```

Comment:
generateEquals may produce UNKNOWN, so we should change it to generateIsTrue(generateEquals).